src/drivers/uavcan/node_info.hpp

```diff
@@ -0,0 +1,166 @@
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#pragma once
+#include <uavcan/protocol/node_info_retriever.hpp>
+#include <uORB/Publication.hpp>
+#include <uORB/topics/device_information.h>
+#include <stdint.h>  // For UINT8_MAX, UINT32_MAX
+#include <px4_platform_common/time.h>
+
+using namespace time_literals;
+
+constexpr int 		DEVICE_INFO_PUBLISH_INTERVAL_MS 	= 1000;
+constexpr uint64_t 	DEVICE_INFO_PUBLISH_RATE_LIMIT_US 	= 100_ms;
+
+class NodeInfoPublisher : private uavcan::INodeInfoListener, private uavcan::TimerBase
+{
+public:
+	enum class DeviceCapability : uint8_t {
+		NONE = UINT8_MAX,  // Invalid/unset capability value (255)
+		GENERIC = 0,
+		AIRSPEED = 1,
+		ESC = 2,
+		SERVO = 3,
+		GPS = 4,
+		MAGNETOMETER = 5,
+		PARACHUTE = 6,
+		RANGEFINDER = 7,
+		WINCH = 8,
+		BAROMETER = 9,
+		OPTICAL_FLOW = 10,
+		ACCELEROMETER = 11,
+		GYROSCOPE = 12,
+		DIFFERENTIAL_PRESSURE = 13,
+		BATTERY = 14,
+		HYGROMETER = 15,
+	};
+
+	NodeInfoPublisher(uavcan::INode &node, uavcan::NodeInfoRetriever &node_info_retriever);
+	~NodeInfoPublisher();
+
+	// Called by sensor bridges to register device capabilities
+	void registerDeviceCapability(uint8_t node_id, uint32_t device_id, DeviceCapability capability);
+
+private:
+	struct NodeInfo {
+		NodeInfo(uavcan::NodeID id, const uavcan::protocol::GetNodeInfo_::Response &node_info)
+			: node_id(id), sw_major(node_info.software_version.major), sw_minor(node_info.software_version.minor),
+			  vcs_commit(node_info.software_version.vcs_commit), hw_major(node_info.hardware_version.major),
+			  hw_minor(node_info.hardware_version.minor)
+		{
+			memcpy(name, node_info.name.c_str(), node_info.name.capacity());
+			name[node_info.name.capacity() - 1] = '\0';
+			memcpy(unique_id, &node_info.hardware_version.unique_id.front(), node_info.hardware_version.unique_id.size());
+		}
+		NodeInfo() = default;
+
+		uavcan::NodeID node_id{};
+
+		char name[uavcan::protocol::GetNodeInfo_::Response::FieldTypes::name::MaxSize];
+		uint8_t unique_id[uavcan::protocol::GetNodeInfo_::Response::FieldTypes::hardware_version::FieldTypes::unique_id::MaxSize];
+		uint8_t sw_major;
+		uint8_t sw_minor;
+		uint32_t vcs_commit;
+		uint8_t hw_major;
+		uint8_t hw_minor;
+	};
+
+	struct DeviceInformation {
+		uint8_t node_id;
+		uint32_t device_id;
+		DeviceCapability capability;
+		bool has_node_info;
+		bool has_capability;
```

Comment:
Why not use `node_id == UINT8_MAX` and `capability == DeviceCapability::None` instead of extra fields?